### ISSUE TITLE
Bug in SpaceTimeNonlocalBlock's softmax operation

### ISSUE THREAD

**Issue by shenshanf** [RaoHaocheng/NLCA_Net#11]
There is an implementation error in the softmax operation of SpaceTimeNonlocalBlock. The current code incorrectly flattens and reshapes the attention matrix before applying softmax.
Current implementation:
```python 
tf.matmul(theta, phi)   
theta = tf.reshape(theta, shape=(batchsize, -1))  # incorrect
theta = tf.nn.softmax(theta, axis=-1)
theta = tf.reshape(theta, shape=(batchsize, imgNum*height*width, imgNum*height*width))
```
The softmax should be applied directly on the last dimension of shape [batch, imgNumheightwidth, imgNumheightwidth]. The reshape operations are unnecessary and incorrect.
Proposed fix:
```python
tf.matmul(theta, phi)   # shape: [batch, imgNum*height*width, imgNum*height*width]
theta = tf.nn.softmax(theta, axis=-1)
```
This matches the standard non-local block implementation where softmax is applied along the attention dimension.

**Comment by shenshanf** [comment]
# Additional note on multiple NonLocal blocks
The paper's Figure 3 shows only one 3D Non-Local block, but the implementation uses three consecutive blocks:
```python
def NonLocalGroupBlock(x, training=True):
    with tf.variable_scope("NonLocalGroupBlock"):
        x = SpaceTimeNonlocalBlock(x, "NonLocal_0", training=training)
        x = SpaceTimeNonlocalBlock(x, "NonLocal_1", training=training)
        x = SpaceTimeNonlocalBlock(x, "NonLocal_2", training=training)
    return x
```
This discrepancy between paper and code is significant given that 3D Non-Local operations are computationally expensive, even at lower resolutions
